### PR TITLE
[Tumblr] Add login support

### PR DESCRIPTION
**Pull request by zackfern**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

I've added login support for Tumblr. Some time in the last year, Tumblr has made it impossible to download explicit content without logging in. When you try to view explicit content without being logged in, you're redirected to a splash screen (https://www.tumblr.com/safe-mode?url=...) which youtube-dl then passes to the generic downloader. This results in HTML pages being downloaded with weird names, such as "safe-mode-168487385934.unknown_video"

The addition of login support allows these URLs to be downloaded without issue. I've tested this locally with both the regular `--username` and `--password` flags and the ~/.netrc file and `-n` flag.

I'm not a Python developer by trade, so please excuse the code if it's a bit off. I believe the code passes flake8 - I ran `flake8 tumblr.py` locally and there wasn't any output.

Please let me know if there's anything else I can do to improve the code. Thanks for your work in maintaining youtube-dl!

### CODE REVIEW

**Comment by Hrxn** [comment]
Nice.

Good idea, relevant improvement, Thanks for your PR.

Although, for the sake of completeness, I want to mention that there is another method to use youtube-dl in this case, as a workaround:

Sign in to your Tumblr account, enable the content settings in your account preferences, and then export the cookies, either manually or by using on of the many browser extensions for that.
Then, running youtube-dl together with `--cookies COOKIEFILE` works as well..

File: youtube_dl/extractor/tumblr.py [hunk 1]
```diff
@@ -97,6 +105,31 @@ class TumblrIE(InfoExtractor):
...
+        webpage = self._download_webpage(self._LOGIN_URL, None, False)
+        form = self._hidden_inputs(webpage)
+        form.update({
+            'user[email]': username,
+            'user[password]': password
+        })
+        post_data = urlencode_postdata(form)
+        login_request = sanitized_Request(self._LOGIN_URL, post_data)
+        login_request.add_header('Content-Type', 'application/x-www-form-urlencoded')
+        login_request.add_header('Referer', self._LOGIN_URL)
```
Comment (dstftw) [inline youtube_dl/extractor/tumblr.py]: Move data and query into `_download_webpage` call.

File: youtube_dl/extractor/tumblr.py [hunk 2]
```diff
@@ -97,6 +105,31 @@ class TumblrIE(InfoExtractor):
...
+        login_response = self._download_webpage(login_request, None, False, 'Wrong login info')
+
+        # Check the login response from Tumblr for an error message and fail the extraction if we find one.
+        login_errors = self._search_regex(r'Tumblr\.RegistrationForm\.errors = \[(.*)\]', login_response, 'login errors', False, False)
```
Comment (dstftw) [inline youtube_dl/extractor/tumblr.py]: 1. Relax whitespace.
2. default and fatal are not used together.
3. Should not match empty string.

File: youtube_dl/extractor/tumblr.py [hunk 3]
```diff
@@ -97,6 +105,31 @@ class TumblrIE(InfoExtractor):
...
+        if login_errors:
+            raise ExtractorError("Error logging in: %s" % login_errors)
```
Comment (dstftw) [inline youtube_dl/extractor/tumblr.py]: 1. Single quotes.
2. `expected`.

File: youtube_dl/extractor/tumblr.py [hunk 4]
```diff
@@ -4,11 +4,19 @@
...
+    int_or_none,
+    sanitized_Request,
+    urlencode_postdata
+)
 
 
 class TumblrIE(InfoExtractor):
     _VALID_URL = r'https?://(?P<blog_name>[^/?#&]+)\.tumblr\.com/(?:post|video)/(?P<id>[0-9]+)(?:$|[/?#])'
+    _NETRC_MACHINE = 'tumblr'
+    _LOGIN_REQUIRED = False
```
Comment (dstftw) [inline youtube_dl/extractor/tumblr.py]: Unused.

File: youtube_dl/extractor/tumblr.py (hunk 2 above)
Comment (zackfern) [reply youtube_dl/extractor/tumblr.py]: - Could you please elaborate on what you mean by "relax whitespace?" Are you saying I should modify the RegEx to ignore whitespace?
- Please correct me if I'm wrong, but I don't think default and fatal are being used together here. The default is set to false and the fatal flag (which is true by default) is set to false. If I shouldn't be setting both options what would you suggest I do instead?

Comment (dstftw) [reply youtube_dl/extractor/tumblr.py]: 1. Regex should match any variation of whitespace around `=`.
2. Once you provide default, fatal is not used thus it's pointless to provide fatal along with default.

**Comment by Hrxn** [comment]
I think this might be related (Improvements to Tumblr extractor): https://github.com/rg3/youtube-dl/issues/14764

Because youtube-dl currently does not get FHD videos, even if available.

I realize that this issue probably looks a lot more verbose and complicated than it actually is, sorry for that 😄 

I searched around a bit, and there are, among others, some userscripts which seem to do the same thing.
For example this one here: https://greasyfork.org/en/scripts/32038-tumblr-hd-video-download-buttons

I think the really relevant part is basically this:

```JavaScript
// Typical Video URL Patterns:
// https://vt.media.tumblr.com/tumblr_ID_NUM.mp4
// https://vtt.tumblr.com/tumblr_ID_NUM.mp4
// https://vt.tumblr.com/tumblr_ID_NUM.mp4

var loc = location.toString();

// ----------------------------------------
// DIRECT MP4 URLS
// ----------------------------------------

function redirectToHD() {
	// Check that the URL is a ~.mp4
	if (!loc.endsWith('.mp4')) return;

	var lowQuality = /[$_]\d*.mp4$/;
	// Do not redirect if already HD
	if (!loc.match(lowQuality)) return;
	// Change to HD
	loc = loc.replace(lowQuality, '.mp4');

	// If the URL is HTTP, change it to HTTPS
	if (!loc.startsWith('https://')) {
		loc = loc.replace(/^http/, 'https');
	}

	// Redirect to the HD video
	location.replace(loc);
}
redirectToHD();

```

**Comment by zackfern** [comment]
Hi @dstftw, I made the changes you requested. Could you please take a look when you have a moment? :) Thanks!

File: youtube_dl/extractor/tumblr.py [hunk 5]
```diff
@@ -97,6 +104,31 @@ class TumblrIE(InfoExtractor):
...
+            return
+        self.report_login()
+        webpage = self._download_webpage(self._LOGIN_URL, None, False)
+        form = self._hidden_inputs(webpage)
+        form.update({
+            'user[email]': username,
+            'user[password]': password
+        })
+        login_response = self._download_webpage(
+            sanitized_Request(self._LOGIN_URL, urlencode_postdata(form), {
```
Comment (dstftw) [inline youtube_dl/extractor/tumblr.py]: Nothing changed.

File: youtube_dl/extractor/tumblr.py [hunk 6]
```diff
@@ -97,6 +104,31 @@ class TumblrIE(InfoExtractor):
...
+                'Content-Type': 'application/x-www-form-urlencoded',
+                'Referer': self._LOGIN_URL
+            }), None, False, 'Wrong login info')
```
Comment (dstftw) [inline youtube_dl/extractor/tumblr.py]: `False` is not a valid note.

File: youtube_dl/extractor/tumblr.py [hunk 7]
```diff
@@ -97,6 +104,31 @@ class TumblrIE(InfoExtractor):
...
 
+    def _real_initialize(self):
+        self._login()
+
+    def _login(self):
+        (username, password) = self._get_login_info()
+        if username is None:
+            return
+        self.report_login()
+        webpage = self._download_webpage(self._LOGIN_URL, None, False)
```
Comment (dstftw) [inline youtube_dl/extractor/tumblr.py]: `False` is not a valid note.

File: youtube_dl/extractor/tumblr.py [hunk 8]
```diff
@@ -97,6 +104,31 @@ class TumblrIE(InfoExtractor):
...
+            'user[password]': password
+        })
+        login_response = self._download_webpage(
+            sanitized_Request(self._LOGIN_URL, urlencode_postdata(form), {
+                'Content-Type': 'application/x-www-form-urlencoded',
+                'Referer': self._LOGIN_URL
+            }), None, False, 'Wrong login info')
+
+        # Check the login response from Tumblr for an error message and fail the extraction if we find one.
+        login_errors = self._search_regex(r'Tumblr\.RegistrationForm\.errors\s*=\s*\[[\"|\'](.+)[\"|\']\]', login_response, 'login errors', False)
```
Comment (dstftw) [inline youtube_dl/extractor/tumblr.py]: `[\"|\']` what is this.

**Pull request closed without merging**